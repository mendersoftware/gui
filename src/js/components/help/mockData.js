// Copyright 2020 Northern.tech AS
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//        http://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.
export const helpProps = {
  docsVersion: undefined,
  isHosted: false,
  menderVersion: 'master',
  isEnterprise: true,
  tenantCapabilities: {
    hasDeviceConfig: true,
    hasMonitor: true
  },
  tokens: [],
  versions: {
    repos: {
      mender: '3.3.0',
      'mender-artifact': '3.8.0',
      'mender-binary-delta': '1.3.1',
      'mender-cli': '1.8.0',
      'mender-configure-module': '1.0.4',
      'mender-connect': '2.0.2',
      'mender-convert': '3.0.0',
      'mender-gateway': '1.0.0',
      'monitor-client': '1.2.0'
    },
    releaseDate: '2003-01-01'
  }
};
