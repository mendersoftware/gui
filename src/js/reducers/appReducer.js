import * as AppConstants from '../constants/appConstants';
import { stringToBoolean } from '../helpers';

const menderEnvironment = {
  hostAddress: null,
  features: {
    hasMultitenancy: false,
    isHosted: false,
    isEnterprise: false,
    isDemoMode: false
  },
  docsVersion: '',
  hostedAnnouncement: '',
  hostedLinks: {},
  menderDebPackageVersion: '',
  integrationVersion: '',
  menderVersion: '',
  menderArtifactVersion: '',
  metaMenderVersion: '',
  services: {
    deploymentsVersion: '',
    deviceauthVersion: '',
    inventoryVersion: '',
    guiVersion: ''
  },
  ...mender_environment
};

const initialState = {
  hostAddress: menderEnvironment.hostAddress,
  snackbar: {
    open: false,
    message: ''
  },
  // return boolean rather than organization details
  features: {
    hasMultitenancy: stringToBoolean(menderEnvironment.features.hasMultitenancy),
    isHosted: stringToBoolean(menderEnvironment.features.isHosted) || window.location.hostname === 'hosted.mender.io',
    isEnterprise: stringToBoolean(menderEnvironment.features.isEnterprise),
    isDemoMode: stringToBoolean(menderEnvironment.isDemoMode)
  },
  hostedAnnouncement: menderEnvironment.hostedAnnouncement,
  docsVersion: isNaN(menderEnvironment.integrationVersion.charAt(0))
    ? ''
    : menderEnvironment.integrationVersion
        .split('.')
        .slice(0, 2)
        .join('.'),
  menderDebPackageVersion: menderEnvironment.menderDebPackageVersion || 'master',
  versionInformation: {
    Integration: isNaN(menderEnvironment.integrationVersion.charAt(0)) ? 'master' : menderEnvironment.integrationVersion,
    'Mender-Client': isNaN(menderEnvironment.menderVersion.charAt(0)) ? 'master' : menderEnvironment.menderVersion,
    'Mender-Artifact': menderEnvironment.menderArtifactVersion,
    'Meta-Mender': menderEnvironment.metaMenderVersion,
    Deployments: menderEnvironment.services.deploymentsVersion,
    Deviceauth: menderEnvironment.services.deviceauthVersion,
    Inventory: menderEnvironment.services.inventoryVersion,
    GUI: menderEnvironment.services.guiVersion || 'latest'
  }
};

const userReducer = (state = initialState, action) => {
  switch (action.type) {
    case AppConstants.SET_SNACKBAR:
      return {
        ...state,
        snackbar: action.snackbar
      };
    case AppConstants.RECEIVED_HOSTED_LINKS:
      return {
        ...state,
        hostedLinks: action.links
      };
    case AppConstants.SET_LOCAL_IPADDRESS:
      return {
        ...state,
        hostAddress: action.ipAddress
      };
    default:
      return state;
  }
};

export default userReducer;
