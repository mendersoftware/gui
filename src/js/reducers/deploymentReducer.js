// Copyright 2019 Northern.tech AS
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//        http://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.
import * as DeploymentConstants from '../constants/deploymentConstants';
import * as DeviceConstants from '../constants/deviceConstants';

export const initialState = {
  byId: {
    // [id]: { statistics: { status: {}, total_size }, devices: { [deploymentId]: { id, log } }, totalDeviceCount }
  },
  byStatus: {
    finished: { deploymentIds: [], total: 0 },
    inprogress: { deploymentIds: [], total: 0 },
    pending: { deploymentIds: [], total: 0 },
    scheduled: { deploymentIds: [], total: 0 }
  },
  config: {
    binaryDelta: {
      timeout: -1,
      duplicatesWindow: -1,
      compressionLevel: -1,
      disableChecksum: false,
      disableDecompression: false,
      inputWindow: -1,
      instructionBuffer: -1,
      sourceWindow: -1
    },
    binaryDeltaLimits: {
      timeout: { ...DeploymentConstants.limitDefault, default: 60, max: 3600, min: 60 },
      sourceWindow: DeploymentConstants.limitDefault,
      inputWindow: DeploymentConstants.limitDefault,
      duplicatesWindow: DeploymentConstants.limitDefault,
      instructionBuffer: DeploymentConstants.limitDefault
    }
  },
  deploymentDeviceLimit: 5000,
  selectedDeviceIds: [],
  selectionState: {
    finished: { ...DeviceConstants.DEVICE_LIST_DEFAULTS, endDate: undefined, search: '', selection: [], startDate: undefined, total: 0, type: '' },
    inprogress: { ...DeviceConstants.DEVICE_LIST_DEFAULTS, perPage: DeploymentConstants.DEFAULT_PENDING_INPROGRESS_COUNT, selection: [] },
    pending: { ...DeviceConstants.DEVICE_LIST_DEFAULTS, perPage: DeploymentConstants.DEFAULT_PENDING_INPROGRESS_COUNT, selection: [] },
    scheduled: { ...DeviceConstants.DEVICE_LIST_DEFAULTS, selection: [] },
    general: {
      state: DeploymentConstants.DEPLOYMENT_ROUTES.active.key,
      showCreationDialog: false,
      showReportDialog: false,
      reportType: null // DeploymentConstants.DEPLOYMENT_TYPES.configuration|DeploymentConstants.DEPLOYMENT_TYPES.software
    },
    selectedId: undefined
  }
};

const deploymentReducer = (state = initialState, action) => {
  switch (action.type) {
    case DeploymentConstants.CREATE_DEPLOYMENT:
      return {
        ...state,
        byId: {
          ...state.byId,
          [action.deploymentId]: {
            ...DeploymentConstants.deploymentPrototype,
            ...action.deployment
          }
        },
        byStatus: {
          ...state.byStatus,
          pending: {
            ...state.byStatus.pending,
            deploymentIds: [...state.byStatus.pending.deploymentIds, action.deploymentId],
            total: state.byStatus.pending.total + 1
          }
        },
        selectionState: {
          ...state.selectionState,
          pending: {
            ...state.selectionState.pending,
            selection: [action.deploymentId, ...state.selectionState.pending.selection],
            total: state.selectionState.pending.total + 1
          }
        }
      };
    case DeploymentConstants.REMOVE_DEPLOYMENT: {
      // eslint-disable-next-line no-unused-vars
      const { [action.deploymentId]: removedDeployment, ...remainder } = state.byId;
      return {
        ...state,
        byId: remainder
      };
    }
    case DeploymentConstants.RECEIVE_DEPLOYMENT:
    case DeploymentConstants.RECEIVE_DEPLOYMENT_DEVICE_LOG:
      return {
        ...state,
        byId: {
          ...state.byId,
          [action.deployment.id]: action.deployment
        }
      };
    case DeploymentConstants.RECEIVE_DEPLOYMENT_DEVICES:
      return {
        ...state,
        byId: {
          ...state.byId,
          [action.deploymentId]: {
            ...state.byId[action.deploymentId],
            devices: action.devices,
            totalDeviceCount: action.totalDeviceCount
          }
        },
        selectedDeviceIds: action.selectedDeviceIds
      };
    case DeploymentConstants.RECEIVE_DEPLOYMENTS:
      return {
        ...state,
        byId: {
          ...state.byId,
          ...action.deployments
        }
      };
    case DeploymentConstants.RECEIVE_INPROGRESS_DEPLOYMENTS:
    case DeploymentConstants.RECEIVE_PENDING_DEPLOYMENTS:
    case DeploymentConstants.RECEIVE_SCHEDULED_DEPLOYMENTS:
    case DeploymentConstants.RECEIVE_FINISHED_DEPLOYMENTS:
      return {
        ...state,
        byStatus: {
          ...state.byStatus,
          [action.status]: {
            ...state.byStatus[action.status],
            deploymentIds: action.deploymentIds,
            total: action.total
          }
        }
      };
    case DeploymentConstants.SELECT_INPROGRESS_DEPLOYMENTS:
    case DeploymentConstants.SELECT_PENDING_DEPLOYMENTS:
    case DeploymentConstants.SELECT_SCHEDULED_DEPLOYMENTS:
    case DeploymentConstants.SELECT_FINISHED_DEPLOYMENTS:
      return {
        ...state,
        selectionState: {
          ...state.selectionState,
          [action.status]: {
            ...state.selectionState[action.status],
            selection: action.deploymentIds,
            total: action.total
          }
        }
      };
    case DeploymentConstants.SET_DEPLOYMENTS_STATE:
      return {
        ...state,
        selectionState: action.state
      };
    case DeploymentConstants.SET_DEPLOYMENTS_CONFIG:
      return {
        ...state,
        config: action.config
      };
    default:
      return state;
  }
};

export default deploymentReducer;
