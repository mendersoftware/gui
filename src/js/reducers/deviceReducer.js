import * as DeviceConstants from '../constants/deviceConstants';

const initialState = {
  byId: {},
  byStatus: {
    [DeviceConstants.DEVICE_STATES.accepted]: { deviceIds: [], total: 0 },
    active: { deviceIds: [], total: 0 },
    inactive: { deviceIds: [], total: 0 },
    [DeviceConstants.DEVICE_STATES.pending]: { deviceIds: [], total: 0 },
    [DeviceConstants.DEVICE_STATES.preauth]: { deviceIds: [], total: 0 },
    [DeviceConstants.DEVICE_STATES.rejected]: { deviceIds: [], total: 0 }
  },
  selectedDevice: null,
  selectedDeviceList: [],
  filters: [
    // { key: 'device_type', value: 'raspberry', operator: '$eq', scope: 'inventory' }
  ],
  filteringAttributes: { identityAttributes: [], inventoryAttributes: [] },
  filteringAttributesLimit: 10,
  total: 0,
  limit: 500,
  groups: {
    byId: {
      [DeviceConstants.UNGROUPED_GROUP.id]: { deviceIds: [], total: 0, filters: [] }
      // dynamo: { deviceIds: [], total: 3, filters: [{ a: 1 }] }
    },
    selectedGroup: null
  }
};

const deviceReducer = (state = initialState, action) => {
  switch (action.type) {
    case DeviceConstants.RECEIVE_GROUPS:
      return {
        ...state,
        groups: {
          ...state.groups,
          byId: { ...state.groups.byId, ...action.groups }
        }
      };
    case DeviceConstants.ADD_TO_GROUP: {
      let group = {
        deviceIds: [action.deviceId],
        filters: [],
        total: 1
      };
      if (state.groups.byId[action.group]) {
        group = {
          filters: [],
          ...state.groups.byId[action.group],
          deviceIds: [...state.groups.byId[action.group].deviceIds, action.deviceId],
          total: state.groups.byId[action.group].total + 1
        };
      }
      return {
        ...state,
        groups: {
          ...state.groups,
          byId: {
            ...state.groups.byId,
            [action.group]: group
          }
        }
      };
    }
    case DeviceConstants.REMOVE_FROM_GROUP: {
      const deviceIdsIndex = state.groups.byId[action.group].deviceIds.findIndex(item => item === action.deviceId);
      const group = {
        ...state.groups.byId[action.group],
        deviceIds: [
          ...state.groups.byId[action.group].deviceIds.slice(0, deviceIdsIndex),
          ...state.groups.byId[action.group].deviceIds.slice(deviceIdsIndex + 1)
        ],
        total: state.groups.byId[action.group].total - 1
      };
      let byId = state.groups.byId;
      let selectedGroup = state.groups.selectedGroup;
      if (group.total || group.deviceIds.length) {
        byId[action.group] = group;
      } else if (state.groups.selectedGroup === action.group) {
        selectedGroup = null;
        delete byId[action.group];
      }
      return {
        ...state,
        groups: {
          ...state.groups,
          byId,
          selectedGroup
        }
      };
    }
    case DeviceConstants.ADD_DYNAMIC_GROUP:
    case DeviceConstants.ADD_STATIC_GROUP:
      return {
        ...state,
        groups: {
          ...state.groups,
          byId: { ...state.groups.byId, [action.groupName]: action.group }
        }
      };
    case DeviceConstants.RECEIVE_DYNAMIC_GROUPS:
      return {
        ...state,
        groups: {
          ...state.groups,
          byId: {
            ...state.groups.byId,
            ...action.groups
          }
        }
      };
    case DeviceConstants.REMOVE_DYNAMIC_GROUP:
    case DeviceConstants.REMOVE_STATIC_GROUP:
      return {
        ...state,
        groups: {
          ...state.groups,
          byId: action.groups
        }
      };
    case DeviceConstants.SELECT_GROUP:
      return {
        ...state,
        selectedDeviceList:
          state.groups.byId[action.group] && state.groups.byId[action.group].deviceIds.length > 0 ? state.groups.byId[action.group].deviceIds : [],
        groups: {
          ...state.groups,
          selectedGroup: action.group
        }
      };
    case DeviceConstants.SELECT_DEVICE:
      return { ...state, selectedDevice: action.deviceId };
    case DeviceConstants.SELECT_DEVICES:
      return { ...state, selectedDeviceList: action.deviceIds };
    case DeviceConstants.RECEIVE_GROUP_DEVICES:
      return {
        ...state,
        groups: {
          ...state.groups,
          byId: {
            ...state.groups.byId,
            [action.groupName]: action.group
          }
        }
      };
    case DeviceConstants.SET_FILTER_ATTRIBUTES:
      return { ...state, filteringAttributes: action.attributes };
    case DeviceConstants.RECEIVE_DEVICE:
      return {
        ...state,
        byId: {
          ...state.byId,
          [action.device.id]: {
            ...state.byId[action.device.id],
            ...action.device
          }
        }
      };
    case DeviceConstants.RECEIVE_DEVICES:
      return {
        ...state,
        byId: {
          ...state.byId,
          ...action.devicesById
        }
      };
    case DeviceConstants.RECEIVE_ALL_DEVICE_IDS: {
      const devicesById = action.deviceIds.reduce((accu, deviceId) => {
        accu[deviceId] = state.byId[deviceId];
        return accu;
      }, {});
      return {
        ...state,
        byId: {
          ...state.byId,
          ...devicesById
        }
      };
    }
    case DeviceConstants.SET_DEVICE_FILTERS: {
      return {
        ...state,
        filters: action.filters
      };
    }

    case DeviceConstants.SET_INACTIVE_DEVICES:
      return {
        ...state,
        byStatus: {
          ...state.byStatus,
          active: {
            deviceIds: action.activeDeviceIds,
            total: action.activeDeviceIds.length
          },
          inactive: {
            deviceIds: action.inactiveDeviceIds,
            total: action.inactiveDeviceIds.length
          }
        }
      };

    case DeviceConstants.SET_PENDING_DEVICES:
    case DeviceConstants.SET_REJECTED_DEVICES:
    case DeviceConstants.SET_PREAUTHORIZED_DEVICES:
    case DeviceConstants.SET_ACCEPTED_DEVICES: {
      const statusDeviceInfo = action.total ? { deviceIds: action.deviceIds, total: action.total } : state.byStatus[action.status];
      return {
        ...state,
        byStatus: {
          ...state.byStatus,
          [action.status]: {
            ...statusDeviceInfo
          }
        }
      };
    }

    case DeviceConstants.SET_UNGROUPED_DEVICES: {
      const ungroupedGroup = action.deviceIds
        ? {
            [DeviceConstants.UNGROUPED_GROUP.id]: {
              ...state.groups.byId[DeviceConstants.UNGROUPED_GROUP.id],
              deviceIds: action.deviceIds,
              total: action.deviceIds.length
            }
          }
        : {};
      return {
        ...state,
        groups: {
          ...state.groups,
          byId: {
            ...state.groups.byId,
            ...ungroupedGroup
          }
        }
      };
    }

    case DeviceConstants.SET_ACCEPTED_DEVICES_COUNT:
    case DeviceConstants.SET_PENDING_DEVICES_COUNT:
    case DeviceConstants.SET_REJECTED_DEVICES_COUNT:
    case DeviceConstants.SET_PREAUTHORIZED_DEVICES_COUNT:
      return {
        ...state,
        byStatus: {
          ...state.byStatus,
          [action.status]: {
            ...state.byStatus[action.status],
            total: action.count
          }
        }
      };
    case DeviceConstants.SET_TOTAL_DEVICES:
      return { ...state, total: action.count };
    case DeviceConstants.SET_DEVICE_LIMIT:
      return { ...state, limit: action.limit };
    case DeviceConstants.RECEIVE_DEVICE_AUTH: {
      const { auth_sets, identity_data, status } = action.device;
      return {
        ...state,
        byId: {
          ...state.byId,
          [action.device.id]: {
            ...state.byId[action.device.id],
            identity_data,
            auth_sets,
            status
          }
        }
      };
    }
    // TODO!!!
    case DeviceConstants.UPDATE_DEVICE_AUTHSET:
    case DeviceConstants.REMOVE_DEVICE_AUTHSET:
    case DeviceConstants.ADD_DEVICE_AUTHSET:
    case DeviceConstants.DECOMMISION_DEVICE:
      return state;
    default:
      return state;
  }
};

export default deviceReducer;
