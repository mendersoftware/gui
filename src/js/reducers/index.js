// Copyright 2019 Northern.tech AS
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//        http://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.
import { combineReducers, configureStore } from '@reduxjs/toolkit';

import { SET_SNACKBAR, UPLOAD_PROGRESS } from '../constants/appConstants';
import { RECEIVE_EXTERNAL_DEVICE_INTEGRATIONS } from '../constants/organizationConstants';
import { USER_LOGOUT } from '../constants/userConstants';
import appReducer from './appReducer';
import deploymentReducer from './deploymentReducer';
import deviceReducer from './deviceReducer';
import monitorReducer from './monitorReducer';
import onboardingReducer from './onboardingReducer';
import organizationReducer from './organizationReducer';
import releaseReducer from './releaseReducer';
import userReducer from './userReducer';

const rootReducer = combineReducers({
  app: appReducer,
  devices: deviceReducer,
  deployments: deploymentReducer,
  monitor: monitorReducer,
  onboarding: onboardingReducer,
  organization: organizationReducer,
  releases: releaseReducer,
  users: userReducer
});

const sessionReducer = (state, action) => {
  if (action.type === USER_LOGOUT) {
    state = undefined;
  }
  return rootReducer(state, action);
};

export const getConfiguredStore = config =>
  configureStore({
    ...config,
    reducer: sessionReducer,
    middleware: getDefaultMiddleware =>
      getDefaultMiddleware({
        immutableCheck: {
          ignoredPaths: ['app.uploadsById']
        },
        serializableCheck: {
          ignoredActions: [RECEIVE_EXTERNAL_DEVICE_INTEGRATIONS, SET_SNACKBAR, UPLOAD_PROGRESS],
          ignoredActionPaths: ['uploads', 'snackbar'],
          ignoredPaths: ['app.uploadsById', 'app.snackbar', 'organization.externalDeviceIntegrations']
        }
      })
  });

export default getConfiguredStore();
