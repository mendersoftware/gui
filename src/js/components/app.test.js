// Copyright 2019 Northern.tech AS
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//        http://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.
import React from 'react';
import { Provider } from 'react-redux';

import { act, screen, waitFor } from '@testing-library/react';
import 'jsdom-worker';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Cookies from 'universal-cookie';

import { defaultState, mockDate, token, undefineds } from '../../../tests/mockData';
import { render } from '../../../tests/setupTests';
import { getConfiguredStore } from '../reducers';
import App, { timeout } from './app';

jest.mock('../tracking');

const mockStore = configureStore([thunk]);

const state = {
  ...defaultState,
  app: {
    ...defaultState.app,
    trackerCode: 'testtracker',
    versionInformation: {
      Integration: 'next'
    }
  },
  deployments: {
    ...defaultState.deployments,
    byId: {},
    byStatus: {
      ...defaultState.deployments.byStatus,
      inprogress: {
        ...defaultState.deployments.byStatus.inprogress,
        total: 0
      }
    },
    deploymentDeviceLimit: null
  },
  users: {
    ...defaultState.users,
    currentUser: null
  }
};

describe('App Component', () => {
  let cookies;
  beforeEach(() => {
    cookies = new Cookies();
    cookies.get.mockReturnValue('omnomnom');
  });
  it('renders correctly', async () => {
    const store = mockStore(state);

    window.localStorage.getItem.mockReturnValueOnce('false');
    const { baseElement } = render(
      <Provider store={store}>
        <App />
      </Provider>
    );
    const view = baseElement;
    expect(view).toMatchSnapshot();
    expect(view).toEqual(expect.not.stringMatching(undefineds));
  });

  it('works as intended', async () => {
    const store = getConfiguredStore({
      preloadedState: {
        ...state,
        users: {
          ...state.users,
          currentUser: 'notNull'
        }
      }
    });
    window.localStorage.getItem.mockReturnValueOnce('false');
    const ui = (
      <Provider store={store}>
        <App />
      </Provider>
    );
    const { rerender } = render(ui);
    cookies.get.mockReturnValue(token);
    act(() => {
      jest.advanceTimersByTime(timeout + 500);
      jest.runAllTicks();
    });
    cookies.get.mockReturnValue('');
    await waitFor(() => rerender(ui));
    act(() => {
      jest.runOnlyPendingTimers();
      jest.runAllTicks();
    });
    await waitFor(() => expect(screen.queryByText(/Version:/i)).not.toBeInTheDocument(), { timeout: 5000 });
    expect(screen.queryByText(/Northern.tech/i)).toBeInTheDocument();
    expect(screen.queryByText(`Â© ${mockDate.getFullYear()} Northern.tech`)).toBeInTheDocument();
  }, 7000);
});
