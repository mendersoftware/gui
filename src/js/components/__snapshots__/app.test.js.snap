// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component renders correctly 1`] = `
Array [
  <div
    className="header "
    id="fixedHeader"
  >
    <div
      className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
      style={
        Object {
          "backgroundColor": "#fff",
          "height": "56px",
          "minHeight": "unset",
          "paddingLeft": "16px",
          "paddingRight": "16px",
        }
      }
    >
      <div
        className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
        style={
          Object {
            "height": "56px",
            "minHeight": "unset",
            "paddingLeft": "16px",
            "paddingRight": "16px",
          }
        }
      >
        <a
          href="/"
          id="logo"
          onClick={[Function]}
        />
      </div>
      <div
        className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
        style={
          Object {
            "flexGrow": "2",
          }
        }
      />
      <div
        className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
        style={
          Object {
            "flexShrink": "0",
          }
        }
      >
        <div>
          <div
            data-for="limit-tip"
            data-offset="{'bottom': 0, 'right': 0}"
            data-tip={true}
            data-tip-disable={false}
            id="limit"
          >
            <div
              className="__react_component_tooltip place-bottom type-dark"
              data-id="tooltip"
              id="limit-tip"
            >
              <h3>
                Device limit
              </h3>
              <p>
                You can still connect another 
                500
                 
                devices
                .
              </p>
              <p>
                Contact us by email at 
                <a
                  href="mailto:support@mender.io"
                >
                  support@mender.io
                </a>
                 to request a higher limit.
              </p>
              <p>
                There is no fee for a higher limit; the purpose of the limit is to allow us to plan capacity for scaling Hosted Mender.
              </p>
              <p>
                Billing is based on your usage only. See
                 
                <a
                  href="https://mender.io/pricing"
                  target="_blank"
                >
                  pricing
                </a>
                 
                for an overview.
              </p>
            </div>
            <div
              className="header-section"
            >
              <a
                className="inline"
                href="/devices"
                onClick={[Function]}
              >
                <span>
                  0
                </span>
                <span>
                  /
                  500
                </span>
                <svg
                  aria-hidden="true"
                  className="MuiSvgIcon-root"
                  focusable="false"
                  role="presentation"
                  style={
                    Object {
                      "fontSize": "20px",
                      "margin": "0 7px 0 10px",
                    }
                  }
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M22 9V7h-2V5c0-1.1-.9-2-2-2H4c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2v-2h2v-2h-2v-2h2v-2h-2V9h2zm-4 10H4V5h14v14zM6 13h5v4H6zm6-6h4v3h-4zM6 7h5v5H6zm6 4h4v6h-4z"
                  />
                </svg>
              </a>
            </div>
          </div>
        </div>
        <a
          className="header-section"
          href="/deployments/"
          onClick={[Function]}
        >
          <span>
            0
          </span>
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root flip-horizontal"
            focusable="false"
            role="presentation"
            style={
              Object {
                "color": "#c7c7c7",
                "fontSize": "20px",
                "margin": "0 7px 0 10px",
                "top": "5px",
              }
            }
            viewBox="0 0 24 24"
          >
            <path
              d="M17.65 6.35C16.2 4.9 14.21 4 12 4c-4.42 0-7.99 3.58-7.99 8s3.57 8 7.99 8c3.73 0 6.84-2.55 7.73-6h-2.08c-.82 2.33-3.04 4-5.65 4-3.31 0-6-2.69-6-6s2.69-6 6-6c1.66 0 3.14.69 4.22 1.78L13 11h7V4l-2.35 2.35z"
            />
          </svg>
        </a>
        <div
          style={
            Object {
              "marginRight": "0",
              "paddingLeft": "30px",
            }
          }
        >
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-text header-dropdown"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            style={
              Object {
                "fill": "rgb(0, 0, 0)",
                "fontSize": "14px",
                "textTransform": "none",
              }
            }
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiButton-label"
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root"
                focusable="false"
                role="presentation"
                style={
                  Object {
                    "color": "#c7c7c7",
                    "fontSize": "20px",
                    "marginRight": "8px",
                    "top": "5px",
                  }
                }
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
                />
              </svg>
              
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </span>
            <span
              className="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="leftFixed leftNav"
  >
    <ul
      className="MuiList-root MuiList-padding"
      style={
        Object {
          "padding": "0",
        }
      }
    >
      <a
        aria-current="page"
        className="MuiListItem-root navLink leftNav MuiListItem-gutters active"
        disabled={false}
        href="/"
        onClick={[Function]}
        style={
          Object {
            "padding": "22px 16px 22px 42px",
          }
        }
      >
        <div
          className="MuiListItemText-root"
          style={
            Object {
              "textTransform": "uppercase",
            }
          }
        >
          <span
            className="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
          >
            Dashboard
          </span>
        </div>
      </a>
      <a
        aria-current={null}
        className="MuiListItem-root navLink leftNav MuiListItem-gutters"
        disabled={false}
        href="/devices"
        onClick={[Function]}
        style={
          Object {
            "padding": "22px 16px 22px 42px",
          }
        }
      >
        <div
          className="MuiListItemText-root"
          style={
            Object {
              "textTransform": "uppercase",
            }
          }
        >
          <span
            className="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
          >
            Devices
          </span>
        </div>
      </a>
      <a
        aria-current={null}
        className="MuiListItem-root navLink leftNav MuiListItem-gutters"
        disabled={false}
        href="/releases"
        onClick={[Function]}
        style={
          Object {
            "padding": "22px 16px 22px 42px",
          }
        }
      >
        <div
          className="MuiListItemText-root"
          style={
            Object {
              "textTransform": "uppercase",
            }
          }
        >
          <span
            className="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
          >
            Releases
          </span>
        </div>
      </a>
      <a
        aria-current={null}
        className="MuiListItem-root navLink leftNav MuiListItem-gutters"
        disabled={false}
        href="/deployments"
        onClick={[Function]}
        style={
          Object {
            "padding": "22px 16px 22px 42px",
          }
        }
      >
        <div
          className="MuiListItemText-root"
          style={
            Object {
              "textTransform": "uppercase",
            }
          }
        >
          <span
            className="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
          >
            Deployments
          </span>
        </div>
      </a>
    </ul>
    <ul
      className="MuiList-root MuiList-padding"
      style={
        Object {
          "bottom": "30px",
          "left": "0px",
          "padding": "0",
          "position": "absolute",
          "right": "0px",
        }
      }
    >
      <a
        className="MuiListItem-root navLink leftNav MuiListItem-gutters"
        disabled={false}
        href="/help"
        onClick={[Function]}
        style={
          Object {
            "padding": "16px 16px 16px 42px",
          }
        }
      >
        <div
          className="MuiListItemText-root"
        >
          <span
            className="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
          >
            Help
          </span>
        </div>
      </a>
      <li
        className="MuiListItem-root MuiListItem-gutters Mui-disabled"
        disabled={true}
        style={
          Object {
            "color": "#949495",
            "padding": "16px 16px 16px 42px",
          }
        }
      >
        <div
          className="MuiListItemText-root MuiListItemText-multiline"
        >
          <span
            className="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
          >
            <div
              aria-describedby={null}
              className=""
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onMouseLeave={[Function]}
              onMouseOver={[Function]}
              onTouchEnd={[Function]}
              onTouchStart={[Function]}
              title={null}
            >
              
            </div>
          </span>
          <p
            className="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary"
          >
            <a
              href="https://docs.mender.io/development/release-information/open-source-licenses"
              rel="noopener noreferrer"
              style={
                Object {
                  "color": "#347A87",
                  "fontSize": "13px",
                  "position": "relative",
                  "top": "6px",
                }
              }
              target="_blank"
            >
              License information
            </a>
          </p>
        </div>
      </li>
    </ul>
  </div>,
  <div
    className="rightFluid container"
  />,
]
`;
